### ISSUE TITLE
Where to find optimal hyperparameters after tuning with ray tune?

### ISSUE THREAD

**Issue by EthanV431** [ultralytics/ultralytics#20103]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

`from ultralytics import YOLO

# Load your model (pretrained weights or config)
model = YOLO("yolo11l.pt")

results = model.tune(
    data="/home/ec2-user/data.yaml",
    epochs=25,                # Reduce epochs per trial
    iterations=100,           # Fewer hyperparameter combinations
    grace_period=5,
    plots=True,
    save=True,
    val=True,
    use_ray=True,
    gpu_per_trial=1,
    fraction=0.1,
    cache=True,             # Resume tuning run
    name="final_tuning_run",           # Directory name where tuning results are saved
    resume=True,
    space={}
)`

I ran the above code in a jupyter notebook and finished a tuning run, but I can't find where the optimal hyperparameters got saved.

All of the data from each iteration got saved into the runs folder in this format:
runs/
└── detect/
    └── final_tuning_run/
        └── _tune_99a5b_00000_0_bgr=0.8523,box=0.1407,cls=1.9553,copy_paste=0.9550,degrees=35.7406,fliplr=0.2641,flipud=0.2525,hsv_h=0.0020,hs_2025-04-09_04-39-19/
            ├── events.out.tfevents.1744173561.ip-10-0-0-252.us-east-2.compute.internal  
            ├── params.json                                                              
            ├── params.pkl                                                               
            ├── progress.csv                                                             
            └── result.json                                                              

There are 100 of these directories, and in each of them the only file I can't open is result.json, which gives me this error when I try:
Renderer Failure: runs/detect/final_tuning_run/_tune_99a5b_00000_0_bgr=0.8523,box=0.1407,cls=1.9553,copy_paste=0.9550,degrees=35.7406,fliplr=0.2641,flipud=0.2525,hsv_h=0.0020,hs_2025-04-09_04-39-19/result.json
Unexpected non-whitespace character after JSON at position 1430 (line 2 column 1)

How can I determine where the best hyperparameters the tuner found without having to go through them one by one? Also, if I have to go through them one by one, how do I determine which were the best ones?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @EthanV431, thank you for your interest in Ultralytics 🚀! We recommend checking out the [Docs](https://docs.ultralytics.com/) for detailed guidance, including [hyperparameter tuning](https://docs.ultralytics.com/guides/hyperparameter-tuning/) and tips for optimizing your model.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. In this case, ensure your code snippet and any relevant log files are included.

If this is a ❓ Question, please share additional details about your environment (e.g., OS, Python version, Ultralytics version) and confirm that you’ve followed the [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

For your specific query regarding the results of hyperparameter tuning with Ray Tune, the results, including the best hyperparameters, should typically be accessible from the tuning run directory. If you’re encountering issues with reading specific files (e.g., `result.json`), provide more details like the error traceback for further assistance. One of our engineers will review your case soon to help clarify the process.

Join the Ultralytics community for additional support:
- For real-time chat, join us on [Discord](https://discord.com/invite/ultralytics) 🎧.
- For in-depth discussions, visit [Discourse](https://community.ultralytics.com/).
- Or share and explore ideas on our [Subreddit](https://reddit.com/r/Ultralytics).

## Upgrade

Please ensure you’re using the latest version of the `ultralytics` package, as updates may resolve your issue. Upgrade with:

```bash
pip install -U ultralytics
```

Verify you are running YOLO in an updated, supported environment with all dependencies installed. YOLO supports the following verified environments:
- **Notebooks**: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM ([Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/))
- **Amazon** Deep Learning AMI ([Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/))
- **Docker Image** ([Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/)) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing, ensuring the stability of YOLO across major platforms.

An Ultralytics engineer will follow up with you shortly. Thanks for your patience and for being a part of the community! 🙌

**Comment by glenn-jocher** [comment]
Hi @EthanV431,

When using Ray Tune with Ultralytics, the best hyperparameters should be saved in a summary file in your tuning directory. After your tuning run completes, look for:

```
runs/detect/final_tuning_run/best_hyperparameters.yaml
```

This file contains the optimal hyperparameters discovered during tuning. It should also include information about which iteration produced the best results and the corresponding fitness metrics.

If that file doesn't exist (which can happen with some Ray Tune configurations), you can programmatically find the best results by:

```python
# At the end of your tuning code
best_result = results.get_best_result(metric="metrics/mAP50-95(B)", mode="max")
print("Best hyperparameters:", best_result.config)
print("Best performance:", best_result.metrics)
```

For manual inspection, the directory names actually contain the hyperparameter values in the format you showed. The best one would be the trial with the highest validation metrics (typically mAP50-95 for detection). You can check the `progress.csv` files in each trial directory to find which one achieved the highest performance.

The Ray Tune integration works differently from the standard Ultralytics tuner, which is why the output format isn't exactly as described in our documentation for the default tuner.
